File: app/src/main/java/com/masterjefferson/flockulator/frameworks/stream/T4JTwitterStreamingApi.java
package com.masterjefferson.flockulator.frameworks.stream;

import com.masterjefferson.flockulator.domain.twitter.auth.data.TwitterCredentialStore;
import com.masterjefferson.flockulator.domain.twitter.auth.model.TwitterCredentials;
import com.masterjefferson.flockulator.domain.twitter.model.BasicTweet;
import com.masterjefferson.flockulator.domain.twitter.stream.data.TwitterStreamingApi;

import io.reactivex.Observable;
import io.reactivex.ObservableEmitter;
import io.reactivex.ObservableOnSubscribe;
import io.reactivex.annotations.NonNull;
import io.reactivex.disposables.Disposable;
import twitter4j.StallWarning;
import twitter4j.Status;
import twitter4j.StatusDeletionNotice;
import twitter4j.StatusListener;
import twitter4j.TwitterStream;
import twitter4j.TwitterStreamFactory;
import twitter4j.conf.Configuration;
import twitter4j.conf.ConfigurationBuilder;

/**
 * ${FILE_NAME}
 * Created by jeff on 9/22/17.
 */

public class T4JTwitterStreamingApi
    implements TwitterStreamingApi {

  private final TwitterCredentialStore credentialStore;
  private       TwitterStream          twitterStream;

  public T4JTwitterStreamingApi(TwitterCredentialStore credentialStore) {
    this.credentialStore = credentialStore;
  }

  @Override
  public Observable<BasicTweet> sample() {
    return Observable.create(new ObservableOnSubscribe<BasicTweet>() {
      @Override
      public void subscribe(@NonNull final ObservableEmitter<BasicTweet> e) throws Exception {
        TwitterStream api = getApiInstance();
        StatusListener listener = new BasicStatusListener() {
          @Override
          public void onStatus(Status status) {
            BasicTweet basicTweet = new BasicTweet(
                status.getId(),
                status.getCreatedAt(),
                status.getText(),
                status.getUser().getName()
            );
            e.onNext(basicTweet);
          }

          @Override
          public void onException(Exception ex) {
            e.onError(ex);
          }
        };
        api.addListener(listener);
        e.setDisposable(new StatusListenerDisposable(api, listener));
        api.sample();
      }
    });
  }

  private TwitterStream getApiInstance() {
    if (twitterStream == null) {
      TwitterCredentials credentials = credentialStore.getCredentials();
      if (credentials == null) {
        throw new NullPointerException("no stored credentials");
      }
      Configuration conf = new ConfigurationBuilder()
          .setOAuthConsumerKey(credentialStore.getConsumerKey())
          .setOAuthConsumerSecret(credentialStore.getConsumerSecret())
          .setOAuthAccessToken(credentials.accessToken)
          .setOAuthAccessTokenSecret(credentials.tokenSecret)
          .build();
      twitterStream = new TwitterStreamFactory(conf).getInstance();
    }
    return twitterStream;
  }

  private static class StatusListenerDisposable
      implements Disposable {

    private final TwitterStream  api;
    private final StatusListener listener;
    private boolean disposed = false;

    private StatusListenerDisposable(TwitterStream api, StatusListener listener) {
      this.api = api;
      this.listener = listener;
    }

    @Override
    public void dispose() {
      api.removeListener(listener);
      disposed = true;
    }

    @Override
    public boolean isDisposed() {
      return disposed;
    }
  }

  private static abstract class BasicStatusListener implements StatusListener {
    @Override
    public void onDeletionNotice(StatusDeletionNotice statusDeletionNotice) {
      //no-op
    }

    @Override
    public void onTrackLimitationNotice(int numberOfLimitedStatuses) {
      //no-op
    }

    @Override
    public void onScrubGeo(long userId, long upToStatusId) {
      //no-op
    }

    @Override
    public void onStallWarning(StallWarning warning) {
      //no-op
    }
  }

}

File: app/src/main/java/com/masterjefferson/flockulator/frameworks/t4j/stream/T4JTwitterStreamingApi.kt
package com.masterjefferson.flockulator.frameworks.stream

import com.masterjefferson.flockulator.domain.twitter.auth.data.TwitterCredentialStore
import com.masterjefferson.flockulator.domain.twitter.model.BasicTweet
import com.masterjefferson.flockulator.domain.twitter.stream.data.TwitterStreamingApi
import com.masterjefferson.flockulator.frameworks.t4j.T4JCompat
import io.reactivex.Observable
import io.reactivex.disposables.Disposable
import io.reactivex.schedulers.Schedulers
import timber.log.Timber
import twitter4j.*
import twitter4j.conf.ConfigurationBuilder
import java.lang.Exception

class StatusListenerDisposable(private val api: TwitterStream,
                               private val listener: StatusListener) : Disposable {

  private var disposed = false

  override fun dispose() {
    T4JCompat.removeStatusListener(api, listener)
    disposed = true
  }

  override fun isDisposed(): Boolean {
    return disposed
  }
}

abstract class BasicStatusListener : StatusListener {
  override fun onTrackLimitationNotice(numberOfLimitedStatuses: Int) {
    Timber.v("track limitation notice: $numberOfLimitedStatuses")
  }

  override fun onStallWarning(warning: StallWarning?) {
    warning?.let { Timber.w("stall warning: $it") }
  }

  override fun onDeletionNotice(statusDeletionNotice: StatusDeletionNotice?) {
//    statusDeletionNotice?.let { Timber.v("deletion notice: $it") }
  }

  override fun onScrubGeo(userId: Long, upToStatusId: Long) {
    Timber.v("scrub geo: $userId, $upToStatusId")
  }
}

fun basicTweet(from: Status): BasicTweet {
  return BasicTweet(from.id, from.createdAt, from.user.name, from.text)
}

class T4JTwitterStreamingApi(private val credentialStore: TwitterCredentialStore) :
    TwitterStreamingApi {

  private var twitterStream: TwitterStream? = null
  private var isRunning = false
  private val apiInstance: TwitterStream
    get() {
      if (twitterStream == null) {
        val confBuilder = ConfigurationBuilder()
            .setAsyncNumThreads(4)
            .setOAuthConsumerKey(credentialStore.consumerKey)
            .setOAuthConsumerSecret(credentialStore.consumerSecret);
        credentialStore.credentials?.let {
          confBuilder.setOAuthAccessToken(it.accessToken)
          confBuilder.setOAuthAccessTokenSecret(it.tokenSecret)
        }
        twitterStream = TwitterStreamFactory(confBuilder.build()).instance
      }
      return twitterStream!!
    }

  override fun sample(): Observable<BasicTweet> {
    val sampleObs: Observable<BasicTweet> = Observable.create { e ->
      val api = apiInstance
      val listener = object : BasicStatusListener() {
        override fun onException(ex: Exception?) {
          ex?.let { e.onError(it) }
        }

        override fun onStatus(status: Status?) {
          status?.let { e.onNext(basicTweet(from = status)) }
        }
      }
      e.setDisposable(StatusListenerDisposable(api, listener))
      T4JCompat.addStatusListener(api, listener)
      if (!isRunning) {
        val query = FilterQuery()
        query.track("taketheknee", "trump", "WhiteHouse", "GOP", "politics", "NFL", "UN", "UnitedNations")
        api.filter(query)
        isRunning = true
      }
    }
    return sampleObs.subscribeOn(Schedulers.io())
  }
}
